pkg/cli/version/version.go

```diff
@@ -1,25 +1,184 @@
 package version
 
 import (
+	"encoding/json"
 	"fmt"
-	"io"
 
 	"github.com/spf13/cobra"
+	"sigs.k8s.io/yaml"
 
-	"k8s.io/apimachinery/pkg/version"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apimachineryversion "k8s.io/apimachinery/pkg/version"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/tools/clientcmd"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	kversion "k8s.io/kubernetes/pkg/kubectl/cmd/version"
+	"k8s.io/kubernetes/pkg/kubectl/util/i18n"
+	"k8s.io/kubernetes/pkg/kubectl/util/templates"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	"github.com/openshift/oc/pkg/version"
 )
 
-// NewCmdVersion provides a shim around version for
-// non-client packages that require version information
-func NewCmdVersion(fullName string, versionInfo version.Info, out io.Writer) *cobra.Command {
+type Version struct {
+	kversion.Version
+	OpenShiftVersion string `json:"openshiftVersion,omitempty"`
+}
+
+var (
+	versionLong = templates.LongDesc(`
+        Print the OpenShift client, Kubernetes server, and OpenShift server versions for the current context.
+        Pass --client to print only the OpenShift client version.`)
+	versionExample = templates.Examples(`
+        # Print the OpenShift client, Kubernetes server, and OpenShift server version information for the current context.
+        %[1]s version
+        # Print the OpenShift client, Kubernetes server, and OpenShift server version numbers for the current context.
+        %[1]s version --short
+        # Print the OpenShift client version information for the current context.
+        %[1]s version --client`)
+)
+
+type VersionOptions struct {
+	oClient         configv1client.ClusterOperatorsGetter
+	discoveryClient discovery.CachedDiscoveryInterface
+
+	genericclioptions.IOStreams
+}
+
+func NewVersionOptions(ioStreams genericclioptions.IOStreams) *VersionOptions {
+	return &VersionOptions{
+		IOStreams: ioStreams,
+	}
+}
+
+// NewCmdVersion is copied from upstream NewCmdVersion with addition of OpenShift Server version info.
+// OpenShift Server version is output only if logged in to a cluster as an admin user.
+func NewCmdVersion(fullName string, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	o := NewVersionOptions(ioStreams)
+	ko := kversion.NewVersionOptions(ioStreams)
```

Comment:
would it be better to embed those options into one struct?

Reply:
_probably_   looking..